rls/src/main/java/io/grpc/rls/internal/CachingRlsLbClient.java

```diff
@@ -0,0 +1,977 @@
+      linkedHashLruCache.close();
+      // TODO(creamsoup) maybe cancel all pending requests
+      pendingCallCache.clear();
+      rlsChannel.shutdown();
+      rlsPicker.close();
+    }
+  }
+
+  /**
+   * Populates async cache entry for new request. This is only methods directly modifies the cache,
+   * any status change is happening via event (async request finished, timed out, etc) in {@link
+   * PendingCacheEntry}, {@link DataCacheEntry} and {@link BackoffCacheEntry}.
+   */
+  private CachedRouteLookupResponse handleNewRequest(RouteLookupRequest request) {
+    synchronized (lock) {
+      PendingCacheEntry pendingEntry = pendingCallCache.get(request);
+      if (pendingEntry != null) {
+        return CachedRouteLookupResponse.pendingResponse(pendingEntry);
+      }
+
+      ListenableFuture<RouteLookupResponse> asyncCall = asyncRlsCall(request);
+      if (!asyncCall.isDone()) {
+        pendingEntry = new PendingCacheEntry(request, asyncCall);
+        pendingCallCache.put(request, pendingEntry);
+        return CachedRouteLookupResponse.pendingResponse(pendingEntry);
+      } else {
+        // async call returned finished future is most likely throttled
+        try {
+          RouteLookupResponse response = asyncCall.get();
+          DataCacheEntry dataEntry = new DataCacheEntry(request, response);
+          linkedHashLruCache.cache(request, dataEntry);
+          return CachedRouteLookupResponse.dataEntry(dataEntry);
+        } catch (Exception e) {
+          BackoffCacheEntry backoffEntry =
+              new BackoffCacheEntry(request, Status.fromThrowable(e), backoffProvider.get());
+          linkedHashLruCache.cache(request, backoffEntry);
+          return CachedRouteLookupResponse.backoffEntry(backoffEntry);
+        }
+      }
+    }
+  }
+
+  public void requestConnection() {
+    rlsChannel.getState(true);
+  }
+
+  /**
+   * Viewer class for cached {@link RouteLookupResponse} and associated {@link ChildPolicyWrapper}.
+   */
+  static final class CachedRouteLookupResponse {
+    private final RouteLookupRequest request;
+
+    // Should only have 1 of following 3 cache entries
+    @Nullable
+    private final DataCacheEntry dataCacheEntry;
+    @Nullable
+    private final PendingCacheEntry pendingCacheEntry;
+    @Nullable
+    private final BackoffCacheEntry backoffCacheEntry;
+
+    CachedRouteLookupResponse(
+        RouteLookupRequest request,
+        DataCacheEntry dataCacheEntry,
+        PendingCacheEntry pendingCacheEntry,
+        BackoffCacheEntry backoffCacheEntry) {
+      this.request = checkNotNull(request, "request");
+      this.dataCacheEntry = dataCacheEntry;
+      this.pendingCacheEntry = pendingCacheEntry;
+      this.backoffCacheEntry = backoffCacheEntry;
+      checkState((dataCacheEntry != null ^ pendingCacheEntry != null ^ backoffCacheEntry != null)
+          && !(dataCacheEntry != null && pendingCacheEntry != null && backoffCacheEntry != null),
```

Comment:
"not all has value" is logically the same as "at least one is no value", but for the latter the expression does not need double negation, sort of a little "simpler".

Reply:
i am not saying it is logically different (i said they are same). 
you can express 2 ways to say at least one value exists using xor. a^b^c except 1 case where all has value. or, you can say a^b^c works when at least one is no value. i found out the 1st is easier to understand because maybe how i memorize the xor truth table. if you says the later one is equally easy to understand i am fine changing it, but just trying to avoid double negation defeat the purpose why it suggest that; it *usually* makes the logic harder to understand.